src/functions.rs

```diff
@@ -2674,6 +2706,165 @@ impl<'a> From<&'a FunctionArgument<'a>> for ExpectedArgInfo<'a, '_> {
     }
 }
 
+// Go through all functions and check if they are castable
+// based only on their args
+pub fn initialize_castable(functions: &mut FunctionMap, types: &TypeMap) {
+    for func in functions.values_mut() {
+        for arg in &func.args {
+            if arg.param_type.is_associated_resource(types) {
+                func.is_castable = false;
+
+                // If we have found one associated resource
+                // we can continue
+                continue;
+            }
+        }
+    }
+}
+
+// Go through all of the functions and check if they are castable
+// base on functions they call.
+pub fn determine_castable(functions: &mut FunctionMap, types: &TypeMap) -> u64 {
+    let mut num_changed: u64 = 0;
+    // We need tmp_functions to avoid a immutable borrow after a mutable one.
+    let tmp_functions = functions.clone();
+    'outer: for func in functions.values_mut() {
+        // If we are already false there is no reason to check our called functions.
+        if !func.is_castable {
+            continue;
+        }
+        for call in func.original_body {
+            if let Statement::Call(call) = call {
+                if let Some(inner_func) = tmp_functions.get(&call.get_cil_name()) {
+                    if !inner_func.is_castable {
+                        num_changed += 1;
+                        func.is_castable = false;
+                        continue 'outer;
+                    }
+                }
+                for arg in &call.args {
+                    if let Argument::Var(arg) = &arg.0 {
+                        // Need to special case this.*
+                        if arg.to_string().contains("this.") {
+                            num_changed += 1;
+                            func.is_castable = false;
+                            continue 'outer;
+                        }
+                        if let Some(ti) = types.get(arg.as_ref()) {
+                            if ti.is_associated_resource(types) {
+                                num_changed += 1;
+                                func.is_castable = false;
+                                continue 'outer;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+    num_changed
+}
+
+pub fn initialize_terminated<'a>(functions: &'a FunctionMap<'a>) -> (Vec<String>, Vec<String>) {
+    let mut term_ret_vec: Vec<String> = Vec::new();
+    let mut nonterm_ret_vec: Vec<String> = Vec::new();
+
+    for func in functions.values() {
+        let mut is_term = true;
+
+        let func_calls = get_all_func_calls(func.original_body.to_vec());
+
+        for call in func_calls {
+            match call.check_builtin() {
```

Comment:
I think we can simplify here with just

```
if call.check_builtin().is_none() {
    is_term=false;
}
```

instead of the full match.  The breaks and continues look unnecessary and the Some(_) case is otherwise empty.